### PR TITLE
opentelemetry-instrumentation-asgi: always set status code on duration attributes

### PR DESCRIPTION
**Pull request by danielhochman**
# Description

Bug: `OpenTelemetryMiddleware` records histograms with no status code if the span is non-recording. The decision to include status code on a histogram should not be dependent on tracing decisions.

## Type of change
- [x] Bug fix (non-breaking change which fixes an issue)

# How Has This Been Tested?
- [x] Unit tests

# Does This PR Require a Core Repo Change?
- [x] No.

# Checklist:
- [x] Followed the style guidelines of this project
- [x] Changelogs have been updated
- [x] Unit tests have been added
- [ ] Documentation has been updated

### CODE REVIEW

**Comment by danielhochman** [comment]
@xrmx Thanks, was waiting on Corp CLA approval. It's signed now!

**Comment by danielhochman** [comment]
@xrmx Thanks again. I think this should be ready to merge!

File: instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py
Comment (xrmx) [inline instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py:713]:
```suggestion
                if message["type"] == "http.response.start":
                    status_code = message["status"]
                    if status_code:
                        duration_attrs[SpanAttributes.HTTP_STATUS_CODE] = (
                            status_code
                        )
                elif message["type"] == "websocket.send":
                    status_code = 200
                    duration_attrs[SpanAttributes.HTTP_STATUS_CODE] = (
                        status_code
                    )
```

What do you think on something like this instead?

Comment (danielhochman) [reply instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py]: Had something like that originally but ran into `R0912: Too many branches (13/12) (too-many-branches)`.

Went back to a clearer if/else and did a slight refactor to pull out some logic to avoid the lint rule.

File: instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py
Comment (xrmx) [inline instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py:718]: Before you were adding the status code only for this two message types and not for everything you get

Comment (danielhochman) [reply instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py:718]: It's effectively the same, right? `status_code` defaults to `None` from :713 unless one of the two message types sets it.

Comment (xrmx) [reply instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py:718]: indeed, sorry for the noise :sweat:

**Comment by danielhochman** [comment]
@xrmx Thanks for your review! Are we ready to merge?

**Comment by lzchen** [comment]
@danielhochman 

Recent changes to asgi instrumentations has caused some conflicts. I will resolve them by directly committing to your branch.

File: instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py
Comment (lzchen) [inline instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py:434]: I don't see much value in making this a helper function.

Comment (danielhochman) [reply instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py:434]: It was either lint ignore too many branches (which I kind of agree with, the function had a lot of difficult to follow indentation) or this helper fn. Let me know which you would prefer and I will make the change.

Comment (lzchen) [reply instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py:434]: I'd add a lint ignore :)

**Review by lzchen — APPROVED** [review]

**Review by xrmx — APPROVED** [review]

**Review by emdneto — APPROVED** [review]
Did som runs in some scenarios I have here and it's working as expected:
For old semconv:
```json
{
    "name": "http.server.duration",
    "description": "Duration of HTTP server requests.",
    "unit": "ms",
    "data": {
        "data_points": [
            {
                "attributes": {
                    "http.scheme": "http",
                    "http.host": "127.0.0.1:8000",
                    "net.host.port": 8000,
                    "http.flavor": "1.1",
                    "http.method": "GET",
                    "http.server_name": "localhost:8000",
                    "http.status_code": 500,
                    "http.target": "/foobar"
                },
            }
```

For new semconv:
```json
{
"name": "http.server.request.duration",
"description": "Duration of HTTP server requests.",
"unit": "s",
"data": {
    "data_points": [
        {
            "attributes": {
                "url.scheme": "http",
                "network.protocol.version": "1.1",
                "http.request.method": "GET",
                "http.route": "/foobar",
                "http.response.status_code": 500,
                "error.type": "500"
}
```

File: instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py
Comment (Kludex) [inline instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py:842]: If anything... It should be 101, not 200.

**Pull request merged**
